tfkbnufft/kbnufft.py

```diff
@@ -170,16 +170,30 @@ def kbnufft_forward_for_interpob(x, om):
         norm = interpob['norm']
         im_rank = interpob.get('im_rank', 2)
 
-        x = scale_and_fft_on_image_volume(
+        fft_x = scale_and_fft_on_image_volume(
             x, scaling_coef, grid_size, im_size, norm, im_rank=im_rank, multiprocessing=multiprocessing)
 
-        y = kbinterp(x, om, interpob)
+        y = kbinterp(fft_x, om, interpob)
 
         def grad(dy):
-            x = adjkbinterp(dy, om, interpob)
-            x = ifft_and_scale_on_gridded_data(
-                x, scaling_coef, grid_size, im_size, norm, im_rank=im_rank)
-            return x, None
+            # Gradients with respect to image
+            grid_dy = adjkbinterp(dy, om, interpob)
+            ifft_dy = ifft_and_scale_on_gridded_data(
+                grid_dy, scaling_coef, grid_size, im_size, norm, im_rank=im_rank)
+
+            # Gradients with respect to trajectory locations
+            r = (
+                tf.linspace(-im_size[0]/2, im_size[0]/2-1, im_size[0]),
+            )
+            if im_rank >=2:
+                r = r + (tf.linspace(-im_size[1]/2, im_size[1]/2-1, im_size[1]),)
+                if im_rank == 3:
+                    r = r + (tf.linspace(-im_size[2]/2, im_size[2]/2-1, im_size[2]),)
+            grid_r = tf.cast(tf.meshgrid(*r, indexing='ij'), x.dtype)[None, ...]
+            fft_dx_dom = scale_and_fft_on_image_volume(
+                x * grid_r, scaling_coef, grid_size, im_size, norm, im_rank=im_rank, multiprocessing=multiprocessing)
```

Comment:
Here multiprocessing shouldn't be taken into account imho because if we are computing gradients we certainly are on GPU
```suggestion
            fft_dx_dom = scale_and_fft_on_image_volume(
                x * grid_r, scaling_coef, grid_size, im_size, norm, im_rank=im_rank)
```

Reply:
Ahh, we cant get gradients on CPU? I was not aware of it..